Prepare UI for attributes configuration

azhiv opened the pull request:
<!---
Copyright (C) 2020-2022 Intel Corporation

SPDX-License-Identifier: MIT
-->

<!-- Raised an issue to propose your change (https://github.com/opencv/cvat/issues).
It helps to avoid duplication of efforts from multiple independent contributors.
Discuss your ideas with maintainers to be sure that changes will be approved and merged.
Read the [CONTRIBUTION](https://github.com/opencv/cvat/blob/develop/CONTRIBUTING.md)
guide. -->

<!-- Provide a general summary of your changes in the Title above -->

### Motivation and context
This PR allows attributes mapping that belong to certain labels (model and task) with ability to pre-map attributes based on their names.
The values returned from nuclio function are also processed during inference to filter out attributes with incompatible types/values.

### How has this been tested?
<!-- Please describe in detail how you tested your changes.
Include details of your testing environment, and the tests you ran to
see how your change affects other areas of the code, etc. -->

### Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply.
If an item isn't applicable by a reason then ~~explicitly strikethrough~~ the whole
line. If you don't do that github will show an incorrect process for the pull request.
If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I submit my changes into the `develop` branch
- [x] I have added a description of my changes into [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
- [ ] I have updated the [documentation](
  https://github.com/opencv/cvat/blob/develop/README.md#documentation) accordingly
- [ ] I have added tests to cover my changes
- [ ] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [x] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2022 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

azhiv commented:
@nmanovic @bsekachev Can you please review the approach for automapping label attributes? This is a follow up PR for functionality developed in scope of https://github.com/openvinotoolkit/cvat/pull/3917
@mikhail-treskin

nmanovic commented:
@azhiv , is it ready? If so, please remove 'draft' mode. Also could you please look at linter issues?

Indeed I believe the problem with copyright header: `Copyright (C) 2019-2021 Intel Corporation`. Need to update the year: `Copyright (C) 2019-2022 Intel Corporation`. If you setup all VS code extensions, it will be updated automatically for files, which you touch. See the list of extensions here: https://openvinotoolkit.github.io/cvat/docs/contributing/development-environment/

![image](https://user-images.githubusercontent.com/40690625/160113740-378f45c4-6851-4524-862e-911c44369040.png)

nmanovic commented:
@klakhov , could you please look at the PR?

azhiv commented:
@klakhov I just finished the implementation, can please trigger the approval checks?

klakhov commented on cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/tools-control.tsx:
```diff
@@ -1022,6 +1022,45 @@ export class ToolsControlComponent extends React.PureComponent<Props, State> {
+
+            switch (functionAttribute.input_type) {
+                case 'number':
+                    return ['select', 'radio', 'text'].includes(dbAttribute.input_type) && !Number.isNaN(value);
```
I believe `dbAttribute` is an instance of `cvat-core` `Attribute` class, so shouln't we check for `dbAttribute.inputType` instead?

azhiv replied:
The object is typed as `any` in the code, so I examined the object structure by debugging the code. The attribute name is snake cased.

azhiv commented on cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/tools-control.tsx:
```diff
@@ -1022,6 +1022,45 @@ export class ToolsControlComponent extends React.PureComponent<Props, State> {
+                return false;
+            }
+
+            if (functionAttribute.input_type === 'checkbox') {
+                return ['true', 'false'].includes(value);
+            }
```
This is an improvement because otherwise this block of code should land in two branches of execution:
1. Inside the if condition on line 1038 where input types are equal.
2. Inside the switch which means the condition for input types equality is not satisfied.
Shifting it up before the condition covers both cases.
Should I add a comment or duplicate this block as described above?

klakhov replied:
I see now. I would suggest to duplicate this code just to improve readability, as it's done to `'number'`

azhiv commented on cvat-ui/src/components/model-runner-modal/detector-runner.tsx:
```diff
@@ -84,14 +96,73 @@ function DetectorRunner(props: Props): JSX.Element {
+            const newMatch: StringObject = {};
+            newMatch[modelAttr] = attrLabel;
+            mapping[modelLabel].attributes = { ...mapping[modelLabel].attributes, ...newMatch };
+            // setMapping({ ...mapping });
```
Thanks for the notice, I'll clean it up.

azhiv commented:
@klakhov A big thanks for your feedback! In fact I missed this case during testing, focusing only on "positive" test cases. I'll fix this issue.
BTW is there a proper place to add tests for this logic?

klakhov commented:
@azhiv Currently we don't test 'AI tools' functionality due to CI reasons, so it's better to check as many cases as possible manually

azhiv commented:
@klakhov I fixed the issue that you pointed out, and added a small usability enhancement (attributes are automapped during labels selection even if the label names differ). Could you please review the changes?

klakhov commented on cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/tools-control.tsx:
```diff
@@ -1022,32 +1022,89 @@ export class ToolsControlComponent extends React.PureComponent<Props, State> {
+                            (data: any): any => {
+                                const jobLabel = jobInstance.labels
+                                    .filter((jLabel: any): boolean => jLabel.name === data.label)[0];
+                                const modelLabel = Object.entries(body.mapping)
+                                    .filter(([, value]) => value === data.label)[0][0];
```
I think this line should be commented, It's not so obvious why we should do `[0][0]`

azhiv replied:
Can I just rewrite the code like so?
```ts
const modelLabel = Object.entries(body.mapping)
    .filter(([, value]) => value === data.label)
    .map(([key]) => key)[0];
```

klakhov commented on cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/tools-control.tsx:
```diff
@@ -1022,32 +1022,89 @@ export class ToolsControlComponent extends React.PureComponent<Props, State> {
+                return ['select', 'radio', 'text'].includes(functionAttribute.input_type);
+            }
+
+            switch (functionAttribute.input_type) {
```
Let's move `functionAttribute.input_type === 'checkbox'` case from if on line 1034 here, some duplications are possible but I think its normal for the sake of readability

azhiv commented:
@klakhov I also noticed this behaviour, but the reason for that is empty attributes array received from the BE
<img width="231" alt="Screenshot 2022-03-31 at 07 08 09" src="https://user-images.githubusercontent.com/32125472/161051004-ae0df003-e4db-439b-adce-22bce2cde27e.png">

I asked @mikhail-treskin, he needs time to investigate this.

klakhov commented:
@azhiv Alright, let us know if you can figure something out :slightly_smiling_face:

azhiv commented:
@klakhov Please help me to understand the requirements.
Here's the label setup I use for testing:
<img width="555" alt="Screenshot 2022-04-07 at 15 40 39" src="https://user-images.githubusercontent.com/32125472/162201271-676eb93a-325b-435b-9758-63fe85747c08.png">

When I run the automatic annotation against the following configuration:
<img width="367" alt="Screenshot 2022-04-07 at 15 41 12" src="https://user-images.githubusercontent.com/32125472/162201506-018c1574-d796-4c7b-9a9c-219dade8a45f.png">

I get the following result:
<img width="297" alt="Screenshot 2022-04-07 at 15 39 22" src="https://user-images.githubusercontent.com/32125472/162201653-8430d3a3-e63a-486e-9c3b-556aa73aaed9.png">

The two non-mapped attributes gain default values which is fine, but the first one's value - Adult - is not in the list of allowed ones - [1, 2, 3].

So the question is - what value is `attribute1` expected to hold in this scenario? The way it works now doesn't seem correct.

klakhov commented:
@azhiv , I would agree with @bsekachev, if we have an attribute value that we are not expected to have, it shouldnt be written into mapped value. It would be great to see warning message about such situation just to let user see what happened.

azhiv commented:
@bsekachev @klakhov Thanks for your feedback! Let's discuss that in more detail.
Should the gained attribute be rejected as a result of `check_attr_value`?

![Screenshot 2022-04-13 at 18 58 20](https://user-images.githubusercontent.com/32125472/163221647-e9421271-0b17-400a-9857-830a18d72a6d.png)

This looks natural to me. However, we should probably enhance the values matching logic in case the attribute types are different but compatible, for example (line 265 in the screenshot) `func_attr_type == "number"` and `db_attr_type == "text"`. Should we convert the value in this case (instead of just examining `value.isnumeric()`)?

merchantbreeze commented:
It's a nice work I need!
It's vaild on the jobs page with AI Tools, but when I used the Automatic annotation with your example model, I've encountered detection error.
![QQ截图20220519223209](https://user-images.githubusercontent.com/29403250/169323060-6a20eed3-1c4c-4b06-a458-5caf28f3430c.png)
![QQ截图20220519223240](https://user-images.githubusercontent.com/29403250/169323166-2b25145d-9c7e-4cd4-9690-cb3c0ab268d0.png)
Here is the logs in cvat
```
2022-05-19 14:32:30,258 DEBG 'rqworker_low' stderr output:
DEBUG - 2022-05-19 14:32:30,258 - worker - Handling failed execution of job 953d039f-1148-44e3-922b-07e8d5e7a576

2022-05-19 14:32:30,260 DEBG 'rqworker_low' stderr output:
ERROR - 2022-05-19 14:32:30,260 - worker - Traceback (most recent call last):
  File "/home/django/cvat/apps/lambda_manager/views.py", line 197, in invoke
    mapped_attributes = attr_mapping[func_label].get("attributes")
KeyError: 'face'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/venv/lib/python3.8/site-packages/rq/worker.py", line 936, in perform_job
    rv = job.perform()
  File "/opt/venv/lib/python3.8/site-packages/rq/job.py", line 684, in perform
    return self.func(*self.args, **self.kwargs)
  File "/home/django/cvat/apps/lambda_manager/views.py", line 618, in __call__
    LambdaJob._call_detector(function, db_task, labels, quality,
  File "/home/django/cvat/apps/lambda_manager/views.py", line 468, in _call_detector
    annotations = function.invoke(db_task, data={
  File "/home/django/cvat/apps/lambda_manager/views.py", line 241, in invoke
    raise ValidationError(
django.core.exceptions.ValidationError: ["`openvino-omz-face-detection-0205` lambda function was called without mandatory argument: 'face'"]
Traceback (most recent call last):
  File "/home/django/cvat/apps/lambda_manager/views.py", line 197, in invoke
    mapped_attributes = attr_mapping[func_label].get("attributes")
KeyError: 'face'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/venv/lib/python3.8/site-packages/rq/worker.py", line 936, in perform_job
    rv = job.perform()
  File "/opt/venv/lib/python3.8/site-packages/rq/job.py", line 684, in perform
    self._result = self._execute()
  File "/opt/venv/lib/python3.8/site-packages/rq/job.py", line 690, in _execute
    return self.func(*self.args, **self.kwargs)
  File "/home/django/cvat/apps/lambda_manager/views.py", line 618, in __call__
    LambdaJob._call_detector(function, db_task, labels, quality,
  File "/home/django/cvat/apps/lambda_manager/views.py", line 468, in _call_detector
    annotations = function.invoke(db_task, data={
  File "/home/django/cvat/apps/lambda_manager/views.py", line 241, in invoke
    raise ValidationError(
django.core.exceptions.ValidationError: ["`openvino-omz-face-detection-0205` lambda function was called without mandatory argument: 'face'"]

2022-05-19 14:32:30,260 DEBG 'rqworker_low' stderr output:
DEBUG - 2022-05-19 14:32:30,260 - worker - Invoking exception handler <function rq_handler at 0x7f446a60bee0>

2022-05-19 14:32:30,273 DEBG 'rqworker_low' stderr output:
DEBUG - 2022-05-19 14:32:30,273 - worker - Sent heartbeat to prevent worker timeout. Next one should arrive within 480 seconds.

2022-05-19 14:32:30,273 DEBG 'rqworker_low' stderr output:
DEBUG - 2022-05-19 14:32:30,273 - worker - Sent heartbeat to prevent worker timeout. Next one should arrive within 480 seconds.

2022-05-19 14:32:30,274 DEBG 'rqworker_low' stderr output:
DEBUG - 2022-05-19 14:32:30,274 - worker - *** Listening on low...

2022-05-19 14:32:30,275 DEBG 'rqworker_low' stderr output:
DEBUG - 2022-05-19 14:32:30,275 - worker - Sent heartbeat to prevent worker timeout. Next one should arrive within 480 seconds.
```
It seems like an error about attribute dict, key 'face' is not exist in the dict so we got a KeyError.

merchantbreeze commented:
I have solved it.I found that the respond method of Automatic Annotation is different with AI Tool in jobs, the attrsMapping is missing from the args, so we can't get it.

bsekachev commented:
Hi @azhiv 

Thank you for the contribution.
We've finished the patch and merged it to our current CVAT fork: https://github.com/cvat-ai/cvat

You can see corresponding commit [here](https://github.com/cvat-ai/cvat/commit/2d522c8781f8de9d06c3a6969482ac7d8c5261e3) :)
